File: python/numerical-stacking-puzzle/define.py
"""
Copyright 2026 Miyabi

Project Name  : Numerical Stacking Puzzle (Python)
File Name     : define.py
Author        : Miyabi
Create        : 2026/06/01
Purpose       : 全ファイル共通宣言
"""

#
# Import
#
from __future__ import annotations

#
# Definition
#

E_APP_NAME_STR = "Numerical Stacking Puzzle"    # アプリケーション名
E_APP_VER_STR  = "Version 0.02"                 # アプリケーションバージョン文字列
E_APP_VER_NUM  = 1                              # アプリケーションバージョン番号
E_APP_CR_STR   = "Copyright(C) 2026 Miyabi."    # アプリケーション著作権情報

E_APP_SAVE_ID_KEY = "N.S.P.SaveData.Version"    # 保存データの識別キー
E_APP_SAVE_ID_NUM = 1                           # 保存データの識別Ver

E_WID_X = 1600
E_WID_Y = 900

#
# 上限値
#
E_CELL_X_NUM = 11   # セル数(横方向)
E_CELL_Y_NUM = 7    # セル数(縦方向)

E_NEXT_BOX_NUM = 4  # 次回箱の数

# 数値評価用マクロ
def numok(n: int, m: int) -> bool:
    return 0 <= n < m

def max2(a: int, b: int) -> int:
    return a if a > b else b

def dec_val(n: int, m: int = 1) -> int:
    # C++ inline Dec(long &n, long m=1)
    return 0 if n <= m else n - m

# 矩形幅算出
def rect_width(rc) -> int:
    return rc[2] - rc[0]

# 矩形高さ算出
def rect_height(rc) -> int:
    return rc[3] - rc[1]

# 環境設定を取得
def get_env(key: str, default: int) -> int:
    return default
# 環境設定を設定
def save_env(key: str, value: int) -> None:
    pass

File: cpp/numerical-stacking-puzzle/define.cpp
/********************************************************************************
* Copyright 2026 Miyabi															*
*																				*
* Project Name	: Numerical Stacking Puzzle (Windows 11, Visual Studio 2022)	*
* File Name		: define.cpp													*
* Author		: Miyabi														*
* Create		: 2026/06/01 													*
* Purpose		: 全ファイル共通宣言											*
********************************************************************************/

/********************************************************************************
*	Include Files																*
********************************************************************************/
#include "framework.h"	/* 必須 */
#include "define.h"
#include "fileio.h"
#include "wmstr.h"

/********************************************************************************
*	Global Functions															*
********************************************************************************/

/********************************************************************************
* FunctionName:	RectWidth														*
* Function    :	矩形幅算出														*
* Argument    :	rc=矩形															*
* Return      :	矩形幅															*
********************************************************************************/
long	RectWidth(RECT &rc)
{
	return rc.right - rc.left;
}

/********************************************************************************
* FunctionName:	RectHeight														*
* Function    :	矩形高さ算出													*
* Argument    :	rc=矩形															*
* Return      :	矩形高さ														*
********************************************************************************/
long	RectHeight(RECT &rc)
{
	return rc.bottom - rc.top;
}

/********************************************************************************
* FunctionName:	GetEnv															*
* Function    :	環境設定を取得													*
* Argument    :	キーコード、取得できなかった場合のデフォルト値					*
* Return      :	環境設定値														*
********************************************************************************/
#define ENV_MAX 4096
WCHAR aEnv[ENV_MAX];

long	GetEnv(const LPWSTR lpKeyCode, long nDef)
{
	long n;
	char aFileBuf[ENV_MAX];

	if(aEnv[0] == WCHAR_NULL)
	{
		// 環境データ無し -> 読み込み
		CFileIO fio;
		if(!fio.Open((LPWSTR)(L"environment.txt"), FOPEN_READ) ||
		   !fio.Read(aFileBuf, Min(fio.Size(), ENV_MAX)))
		{
			// 読み込み失敗
			aEnv[1] = WCHAR_NULL;
			aEnv[0] = WCHAR_NULL;
			return nDef;
		}
		// 読み込み成功
		aFileBuf[Min(fio.Size(), ENV_MAX)-1] = '\0';
		mbstowcs_s(nullptr, aEnv, ENV_MAX, aFileBuf, ENV_MAX);
		aEnv[ENV_MAX-1] = WCHAR_NULL;
		_wcsupr_s(aEnv, ENV_MAX);	// 大文字に統一する
		for(n = (long)wcslen(aEnv); n >= 0; n--)
		{
			if(aEnv[n] == L'=')
				aEnv[n] = L' ';
		}
	}
	if(wcslen(aEnv) <= 3)
	{
		// 有効な環境データがないためデフォルト値を返す
		return nDef;
	}

	// キーコード取得
	CWMStr wmsEnv;
	wmsEnv.SetW(aEnv);
	wmsEnv.SeekClrW();
	if(wmsEnv.SeekFindW(lpKeyCode))
	{
		wmsEnv.SeekMoveW(wcslen(lpKeyCode));
		if(wmsEnv.SeekScanDigW(&n) == 0)
		{
			// 有効な環境データがないためデフォルト値を返す
			return nDef;
		}
		return n;
	}

	// 有効な環境データがないためデフォルト値を返す
	return nDef;
}

/********************************************************************************
* FunctionName:	SaveEnv															*
* Function    :	環境設定を設定													*
* Argument    :	キーコード、設定値												*
* Return      :	none															*
********************************************************************************/
void	SaveEnv(const LPWSTR lpKeyCode, long nVal)
{
	static bool bFirst = FALSE;
	CFileIO fio;
	if(!fio.Open((LPWSTR)(L"environment.txt"), (bFirst) ? FOPEN_APPEND : FOPEN_WRITE))
		return;

	CWMStr wmsFileBuf;

	wmsFileBuf.SetW(lpKeyCode);
	wmsFileBuf.AddW((LPWSTR)(L"="));
	if(!fio.WriteLn(wmsFileBuf.PtW(), nVal))
		return;

	bFirst = TRUE;
}
